include/kernels/unitarize_force.cuh

```diff
@@ -0,0 +1,362 @@
+#pragma once
+
+#include <quda_matrix.h>
+#include <gauge_field_order.h>
+#include <color_spinor.h>
+#include <svd_quda.h>
+#include <kernel.h>
+
+namespace quda {
+
+  static double unitarize_eps;
+  static double force_filter;
+  static double max_det_error;
+  static bool   allow_svd;
+  static bool   svd_only;
+  static double svd_rel_error;
+  static double svd_abs_error;
+
+  namespace fermion_force {
+
+    template <typename Float_, int nColor_, QudaReconstructType recon_, QudaGaugeFieldOrder order = QUDA_NATIVE_GAUGE_ORDER>
+    struct UnitarizeForceArg : kernel_param<> {
+      using Float = Float_;
+      static constexpr int nColor = nColor_;
+      static constexpr QudaReconstructType recon = recon_;
+      // use long form here to allow specification of order
+      typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_NO,2*nColor*nColor,QUDA_STAGGERED_PHASE_NO,gauge::default_huge_alloc, QUDA_GHOST_EXCHANGE_INVALID,false,order>::type F;
+      typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_NO,2*nColor*nColor,QUDA_STAGGERED_PHASE_NO,gauge::default_huge_alloc, QUDA_GHOST_EXCHANGE_INVALID,false,order>::type G;
+      F force;
+      const F force_old;
+      const G u;
+      int *fails;
+      const double unitarize_eps;
+      const double force_filter;
+      const double max_det_error;
+      const int allow_svd;
+      const int svd_only;
+      const double svd_rel_error;
+      const double svd_abs_error;
+
+      UnitarizeForceArg(GaugeField &force, const GaugeField &force_old, const GaugeField &u, int *fails,
+			double unitarize_eps, double force_filter, double max_det_error, int allow_svd,
+			int svd_only, double svd_rel_error, double svd_abs_error) :
+        kernel_param(dim3(u.VolumeCB(), 2, 1)),
+        force(force),
+        force_old(force_old),
+        u(u),
+        fails(fails),
+        unitarize_eps(unitarize_eps),
+        force_filter(force_filter),
+        max_det_error(max_det_error),
+        allow_svd(allow_svd),
+        svd_only(svd_only),
+        svd_rel_error(svd_rel_error),
+        svd_abs_error(svd_abs_error) { }
+    };
+
+    template <class Real> class DerivativeCoefficients {
+      Real b[6];
+      constexpr Real computeC00(const Real &u, const Real &v, const Real &w)
+      {
+        return -pow(w,3) * pow(u,6) + 3*v*pow(w,3)*pow(u,4) + 3*pow(v,4)*w*pow(u,4)
```

Comment:
Use `math_helper` specialization for integer powers

Reply:
a9c7cbdeb7aed146422c0f7333c2ee532c2a1c20.  Also improved the associated the unit test for this to remove dead code and use a tighter tolerance when running in double precision.